taichi/transforms/lower_ast.cpp

```diff
@@ -167,11 +169,67 @@ class LowerAST : public IRVisitor {
       auto end = stmt->end;
       begin->flatten(flattened);
       end->flatten(flattened);
-      auto &&new_for = std::make_unique<RangeForStmt>(
-          stmt->parent->lookup_var(stmt->loop_var_id[0]), begin->stmt,
-          end->stmt, std::move(stmt->body), stmt->vectorize, stmt->parallelize,
-          stmt->block_dim, stmt->strictly_serialized);
-      flattened.push_back(std::move(new_for));
+      bool is_good_range_for = capturing_loop == nullptr || // the outer-most for
+        std::find(detected_fors_with_break.begin(), detected_fors_with_break.end(), stmt)
+        == detected_fors_with_break.end(); // for without break
+      if (is_good_range_for) {  // #578
+        auto &&new_for = std::make_unique<RangeForStmt>(
+            stmt->parent->lookup_var(stmt->loop_var_id[0]), begin->stmt,
+            end->stmt, std::move(stmt->body), stmt->vectorize,
+            stmt->parallelize, stmt->block_dim, stmt->strictly_serialized);
+        flattened.push_back(std::move(new_for));
+      } else {
+        // transform into a structure as
+        // i = begin; while (1) { if (i > end) break; original body; i += 1; }
+        // auto loop_var = Stmt::make<AllocaStmt>(DataType::i32);
+        auto loop_var = stmt->parent->lookup_var(stmt->loop_var_id[0]);
+        flattened.push_back(Stmt::make<LocalStoreStmt>(loop_var, begin->stmt));
+        auto loop_var_addr = LaneAttribute<LocalAddress>(
+            LocalAddress(loop_var->as<AllocaStmt>(), 0));
+        VecStatement flattened2;
+        // auto loop_var_store = Stmt::make<LocalStoreStmt>(loop_var_addr,
+        // begin->stmt);
+        flattened2.push_back(Stmt::make<LocalLoadStmt>(loop_var_addr));
+        auto loop_var_load_stmt = flattened2.back().get();
+        flattened2.push_back(Stmt::make<BinaryOpStmt>(
+            BinaryOpType::cmp_lt, flattened2.back().get(), end->stmt));
+        auto cond_stmt = flattened2.back().get();
+
+        auto &&new_while = std::make_unique<WhileStmt>(std::move(stmt->body));
+        auto mask = std::make_unique<AllocaStmt>(DataType::i32);
+        new_while->mask = mask.get();
+        auto &stmts = new_while->body;
+        for (int i = 0; i < (int)flattened2.size(); i++) {
+          stmts->insert(std::move(flattened2[i]), i);
+        }
+
+        VecStatement flattened3;
+        flattened3.push_back(Stmt::make<ConstStmt>(TypedConstant((int32)1)));
```

Comment:
You can actually use `auto const_one = flattened3.push_back<ConstStmt>(TypedConstant((int32)1)`
Then you don't need `Stmt::make` and `flattened3.back().get()` :-)